servicetalk-transport-netty-internal/src/main/java/io/servicetalk/transport/netty/internal/WriteStreamSubscriber.java

```diff
@@ -420,6 +480,39 @@ private void terminateSubscriber(@Nullable Throwable cause) {
             }
         }
 
+        private void notifyAllListeners(@Nullable Throwable cause) {
+            final ChannelFuture future = cause == null ? channel.newSucceededFuture() : channel.newFailedFuture(cause);
+            while (!listenersOnWriteBoundaries.isEmpty()) {
+                Object mayBeListener = listenersOnWriteBoundaries.removeFirst();
+                if (mayBeListener != WRITE_BOUNDARY) {
+                    notifyListener(future, mayBeListener);
+                }
+            }
+        }
+
+        private void notifyListenersTillNextWrite(@Nullable Throwable cause) {
+            assert !listenersOnWriteBoundaries.isEmpty();
+            Object shdBeWriteBoundary = listenersOnWriteBoundaries.removeFirst();
+            if (shdBeWriteBoundary != WRITE_BOUNDARY) {
+                return;
+            }
+            final ChannelFuture future = cause == null ? channel.newSucceededFuture() : channel.newFailedFuture(cause);
+            while (!listenersOnWriteBoundaries.isEmpty() && listenersOnWriteBoundaries.peekFirst() != WRITE_BOUNDARY) {
+                Object shdBeListener = listenersOnWriteBoundaries.removeFirst();
+                notifyListener(future, shdBeListener);
+            }
+        }
+
+        @SuppressWarnings({"rawtypes", "unchecked"})
+        private void notifyListener(final ChannelFuture future, final Object shdBeListener) {
```

Comment:
nit: consider making this method `static`

Reply:
Actually, this made me notice that I can use `DefaultPromise.notifyListener()`  which provides stack overflow protection 🎉